fix(compiler): improve error message for references to default exports

chuckjaz opened the pull request:
Fixes #12020
Fixes #11034

**What kind of change does this PR introduce?** (check one with "x")
```
[x] Bugfix
```

**What is the current behavior?** (You can also link to an open issue here)

The error message produced when references to module instances and default exports are cryptic.

**What is the new behavior?**

The error messages are more descriptive of the actual error.


**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[x] No
```

chuckjaz committed:
fix(compiler): improve error message for references to default exports

Fixes #12020
Fixes #11034

alfaproject commented on modules/@angular/compiler/test/aot/static_symbol_resolver_spec.ts:
```diff
@@ -235,16 +235,20 @@ describe('StaticSymbolResolver', () => {
   });
 
-  it('should ignore module references without a name', () => {
+  it('should report an error for references to defautl exports', () => {
```
default

chuckjaz replied:
Thanks!

tbosch commented on modules/@angular/compiler/test/aot/compiler_spec.ts:
```diff
@@ -67,6 +67,69 @@ describe('compiler (unbundled Angular)', () => {
              .toBeDefined();
        })));
+
+    it('should be report references to default exports', async(() => {
```
typo: `be report`...

matsko committed:
feat(animations): introduce a wave of new animation features

matsko committed:
refactor(animations): single animation engine code pass

vicb committed:
refactor(router): cleanup, simplification

wardbell committed:
fix(aio): do not nav to an image url

closes #16608
Formerly, tried to navigate when user clicked an anchor with an image url (to view image in a new tab) resulting in 404.
Now ignores href URL with any extension and lets browser handle it.

wardbell committed:
feat(aio): scroll back to top when TOC more-items clicked to close

closes  #16482
Usability feature. When long embedded TOC, collapsing it should scroll user to TOC top, not leave in the middle of the page.

wardbell committed:
refactor(aio): simplify/clarify some observables

petebacondarwin committed:
build(aio): local assets are not dangling links

Closes #16615
Closes #16616

wardbell committed:
refactor(aio): rename auto-scroll -> scroll and refactor

Refactor inspired by @gkalpak feedback on PR #16619

wardbell committed:
docs(aio): doc guide tweaks (GS thru Components)

Working through obvious non-image content defects, from top of menu (Getting Started) through Fundamentals/Components.

petebacondarwin committed:
build(aio): run docs after plunkers and zips (in setup)

Foxandxss committed:
docs(aio): more content fixes

wardbell committed:
feat(aio): scrolling tweaks per 16619 (#16660)

petebacondarwin committed:
build(aio): remove "pure" property from pipe template (#16655)

This information is not relevant to users, right now.

Closes #16641

petebacondarwin committed:
docs(core): remove link to non-existent class in `InjectableDecorator` (#16653)

Closes #16640

petebacondarwin committed:
docs(*) fix dangling links in API docs (#16632)

* docs(animations): fix links to `Component` animations

* docs(core): fix links to `ReflectiveInjector` methods

The `resolve` and other methods were moved from the
`Injector` to the `ReflectiveInjector`.

* docs(core): fix links to `Renderer`

The local links were assuming that that methods were on the
current document (e.g. `RootRenderer`), but they are actually
on the `Renderer` class.

* docs(router): fix links to methods

* docs(forms): fix links to methods

* docs(core): fix links to methods

* docs(router): fix API page links and an internal link

petebacondarwin committed:
docs(aio): image sweep (#16609)

* fix(aio): allow code blocks to clear floated images

Previously the negative margin on the code headings were causing
floated images to overlay the start of a code block. Now all code block
successfully clear all floated elements.

* feat(aio): add a `.clear` class for clearing floating images

* fix(aio): tidy up image styles

The css rules for `img.right` and `img.left` allow authors easy
access to floating an image on the left or right, respectively.

The `.image-display` rule which was always found on a figure
has been simplified so that all figures have this styling. It is very
unlikely that a figure will be used outside the content area; and
at this time it seems like `figure` is as good an indicator that we
want this kind of styling as anything.

Now that images are all tagged with width and height values, we cannot
assume to modify these dimensions via CSS as it can cause the image to
lose its correct proportions.  Until we find a better solition we must set
`height` to `auto` when the screen width is below 1300px to ensure that
these images maintain their proportions as they get shrunk to fit.

* docs(aio): general tidy up of image HTML in guides

Previously, the guides have a lot of inline image styling and unnecessary
use of the `image-display` css class.
Images over 700px are problematic for guide docs, so those have been given
specific widths and associated heights.

* docs(aio): use correct anchor for "back to the top" link

The `#toc` anchor does not work when the page is
wide enough that the TOC is floating to the side.

* build(aio): add `#top-of-page` to path variants for link checking

Since the `#top-of-page` is outside the rendered docs
the `checkAnchorLinks` processor doesn't find them
as valid targets for links.
Adding them as a `pathVariant` solves this problem
but will still catch links to docs that do not actually exist.

* fix(aio): ensure that headings clear floated images

* fix(aio): do not force live-example embedded image to 100% size

This made them look too big, generally. Leaving them with no size means
that they will look reasonable in large viewports and switch to 100% width
in narrow viewports.

petebacondarwin committed:
fix(aio): basic fix to the TOC styling

Without any major refactoring these changes bring the TOC much closer to
how the Google developer docs TOC looks.

Closes #16646

chuckjaz committed:
refactor(compiler-cli): move the expression expression type checker (#16562)

The expression type checker moved from the language service
to the compiler-cli in preparation to using it to check
template expressions.

petebacondarwin committed:
fix(aio): remove unwanted code styling from styleguide doc

Closes #16514

jasonaden committed:
build: update key for pushing to PACKAGE-builds repos (#16667)

sjtrimble committed:
feat(aio): update code style

petebacondarwin committed:
docs(aio): remove excess h1s from guides

Only one h1 is allowed per document.

(Also took the opportunity to remove unnecessary blank lines from these
docs, and a bit of general tidying.)

Closes #16193

petebacondarwin committed:
build(aio): fail the doc-gen if there is an invalid `{@link ...}` tag

This fail behaviour is only turned on for `yarn docs`;
in `yarn docs-watch` you only receive a warning.
This is because you can get false errors when watching
since we don't parse all the docs in that case.

petebacondarwin committed:
build(aio): fail build if more than h1 is found in a doc

Closes #16193

petebacondarwin committed:
docs(common): remove h1s from API docs

Only one h1 is allowed per document, and this is provided by the template.
So we cannot have any h1 tags (or `#` markdown shorthand) in any API docs.

Closes #16193

Foxandxss committed:
docs(aio): add a progress reporter for the webpack guide

Foxandxss committed:
build(aio): make less verbose tools

gkalpak committed:
refactor(aio): remove redundant styles

Due to the new structure of `live-example`, the only style in
`_live-example.scss` was not applied. This is OK according to
https://github.com/angular/angular/issues/15935#issuecomment-295482503, since we
want it to take the full width.

Fixes #15935

Foxandxss committed:
docs(aio): remove warning when testing in the webpack guide

gkalpak committed:
fix(aio): style TOC scrollbar similar to other scrollbars

gkalpak committed:
fix(aio): style sidenav-content scrollbar similar to other scrollbars

petebacondarwin committed:
test(aio): add tests for renderAttributes helper

petebacondarwin committed:
build(aio): automatically link code blocks to API docs

gkalpak committed:
build(aio): serve gzipped content from the preview server

Fixes #16699

jasonaden committed:
release: cut the 4.2.0-beta.1 release

jasonaden committed:
docs: add changelog for 4.2.0-beta.1

aaronhoffman committed:
docs(router): Change CanDeactivate to CanLoad (#16237)

fix mistake in docs. CanDeactivate should be CanLoad

chuckjaz committed:
feat(compiler): support a non-null postfix assert (#16672)

Template expressions can now use a post-fix `!` operator
that asserts the target of the operator is not null. This is
similar to the TypeScript non-null assert operator. Expressions
generated in factories will be generated with the non-null assert
operator.

Closes: #10855

tbosch committed:
refactor: use view engine also for `NgModuleFactory`s (#16658)

* refactor(core): provide error message in stack for reflective DI

Fixes #16355

* fix(compiler): make AOT work with `noUnusedParameters`

Fixes #15532

* refactor: use view engine also for `NgModuleFactory`s

This is a prerequisite for being able to mock providers
in AOTed code later on.

tbosch committed:
refactor(compiler): produce more dense generated code (#16666)

This changes the formatting to be less verbose but still
tries to be readable.

Foxandxss committed:
docs(aio): resize images

jasonaden committed:
docs: add saved replies (#16726)

alxhub committed:
fix(aio): switch to ngu-sw-manifest gen to unblock critical path (#16709)

Jacksunwei committed:
fix(upgrade): use quote to prevent ClossureCompiler obfuscating $event. (#16724)

This is critical for AngularJS to get the $event object in template.

gkalpak committed:
ci(aio): create previews for all PRs that touch non-spec files in `aio/` or `packages/` (#16733)

Previously, no previews would be deployed for PRs that didn't touch files inside
`aio/`. Now, previews will be deployed for PRs that touch non-spec files inside
either `aio/` or `packages/` (as long as other preconditions are met).

Partially addresses #16526.

gkalpak committed:
ci(aio): deploy previews early (right after build) (#16734)

Previously, PR previews were deployed after successfully running all tests.
While this makes sense for staging/production deployments, previews should be
available as soon as possible (and regardless of the outcome of tests).

Fixes #16705

gkalpak committed:
ci: fix syntax error in `deploy.sh` (#16739)

jasonaden committed:
docs: remove angular.io from changelog (#16723)

gkalpak committed:
fix(aio): only register ServiceWorker in production mode

Since abb36e3cb, we no longer rely on the cli to set up ServiceWorker, but do it
manually as part of `yarn build`. When using `ng serve`, registering the
ServiceWorker fails, because we haven't created `ngsw-manifest.json` nor copied
`worker-basic.min.js` into dist.

This commit works around this, by only registering the service worker in
production mode (which is what the cli does too).

Caveat:
It is not possible to enable ServiceWorker with `ng serve`/`yarn start` and
using the `--prod` flag will try to register it, but fail because the necessary
files (`ngsw-manifest.json` and `worker-basic.min.js`) will not be available.
(As a work-around, you can use `yarn build` and serve the files in `dist/` with
`yarn http-server -- dist -p 4200`.)

gkalpak committed:
ci(aio): correctly catch PR preview pre-verification errors

Previously, `aio/aio-builds-setup/scripts/travis-preverify-pr.sh` was supposed
to exit with 1 if a PR did not meet the preconditions and 2 if an error occurred
during pre-verification.
It relied on the exit codes of the node script that did the actual work, but
didn't account for errors that would be thrown in the `sh` script itself (e.g.
if the node script was not available). This caused such errors to appear as
non-verified PRs, instead of real errors that should fail the build.

This commit swaps the exit codes, so that now a 2 means non-verified PR and 1
designates an error.

gkalpak committed:
ci(aio): build `aio-builds-setup` scripts before pre-verifying PR

(Coincidentally), this wasn't an issue before fdfeaaf1f, because
pre-verification was run after `test.sh`, during which `aio-builds-setup` was
built.
Now that `deploy-staging.sh` is being run before `test.sh`, we need to build
the `aio-builds-setup` scripts first.

gkalpak committed:
refactor(aio): drop `run` (from `yarn run`), rename script and remove unnecessary cmds

petebacondarwin committed:
refactor(aio): remove unused mock method

petebacondarwin committed:
fix(aio): set focus back to search box on "escape"

For improved accessibility, the focus should be sent
back to the input box when we are displaying the
search results and the user hits ESC.

See #16005

lozanol committed:
fix(upgrade): Prevent renaming of $inject property (#16706)

Use bracket notation to access $inject in downgradeInjectable to
support property renaming. Since the return type is any,
Closure compiler renames $inject.

vicb committed:
fix(router): Wrap Promise-like instances in native Promises (#16759)

Hybrid apps (mix of Angular and AngularJS) might return AngularJS implementation
of Promises that do not play well with the change detection. Wrapping them in
native Promises fix this issue.

This could be the case when a Resolver returns a `$q` promise.

gkalpak committed:
build: use subshells when changing directories

This prevents being left in the wrong directory in case of error.

IgorMinar committed:
build(aio): upgrade @angular/* and friends, remove obsolete deps

matsko committed:
feat(core): introduce fixture.whenRenderingDone for testing (#16732)

IgorMinar committed:
build(aio): add ngo + purify

matsko committed:
refactor(animations): make animation testing work with fixture.whenRenderingDone

petebacondarwin committed:
refactor(aio): centralise search ui management

Previously the logic for deciding when to display
the search result was spread between different
parts of the application and used non-intuitive logic
such as sending a blank results set to the searchResults.

This commit moves the management of displaying
the search results (and also setting focus of the
search input box) to the AppComponent. This makes
it easier to understand what happens and why; but
also allows the search UI components to be more
easily reused (such as embedding them in the 404
page).

gkalpak committed:
fix(aio): fix PWA testing in cases where `atob`/`btoa` is necessary

In some cases (unclear when), traceviewer-js, used by Lighthouse under the hood,
assumes `atob`/`btoa` are defined in the global scope. This is true for browser
environments, but not on node.

As a result, some aggregations that required access to model-tracing failed to
produce results, dropping the overall PWA score.

This affected #16665 (e.g. commit 0de6eec7a).

gkalpak committed:
ci(aio): log the full PWA testing results

This will help diagnose errors or regressions in PWA scores.

wardbell committed:
test(aio): remove compileComponents from all tests

By reflex we began all component tests with an async `beforeEach` that called `compileComponents`.
In at least one case (`live-example.component.spec.ts`) that led to the `it` tests being async as well.
There is no need to call `.compileComponents` because CLI web pack + plugin inlines all templates and styles.
While `.compileComponents` was harmless, it added complexity and distraction which we should not inflict on future readers and testers.

wardbell committed:
docs(aio): remove fundamentals/techniques intro pages

wardbell committed:
fix(aio): remove searchbox shrink animation

wardbell committed:
fix(aio): no white band below toolbar when mobile

At 600px wide, the buttons on tool bar shrank from 64 to 56px.
Should shrink the content’s top padding from 64 to 56 as well.

mhevery committed:
fix: add typescript 2.3.2 typings test (#16738)

Closes #16663

Adol1111 committed:
fix(compiler-cli): import routing module with forRoot (#16438)

tbosch committed:
feat: introduce `TestBed.overrideProvider` (#16725)

This allows to overwrite all providers for a token, not matter
where they were defined.

This can be used to test JIT and AOT’ed code in the same way.

Design doc: https://docs.google.com/document/d/1VmTkz0EbEVSWfEEWEvQ5sXyQXSCvtMOw4t7pKU-jOwc/edit?usp=sharing

petebacondarwin committed:
fix(aio): support IE via conditionally loaded polyfills

Closes #16519

Foxandxss committed:
docs(aio): fix toh e2e test

wardbell committed:
feat(aio): sidenav headers should be focusable buttons

Sidenav headers had been anchors w/o hrefs. These can’t take focus which makes you can’t navigate through them with keyboard. For a11y purposes, this PR turns them into buttons.

wardbell committed:
feat(aio): code copy button has tooltip and aria-label

Based on optional title passed in from parent element such as CodeExample or CodeTabs.
Darkens uncovered copy button slightly as recommended for a11y.
PR #16808

IgorMinar committed:
perf(aio): load reflect-metadata polyfill only in the dev mode

IgorMinar committed:
build(aio): upgrade to @angular/cli@1.1.0-rc.0

gkalpak committed:
fix(aio): fix PWA testing on Windows (and reduce flaky-ness)

Adding a delay after Lighthouse has run and before killing the Chrome process
avoids `ECONNREFUSED` errors on Windows and will hopefully prevent such errors
occasionally appearing on CI.

Based on info in:
https://github.com/paulirish/pwmetrics/issues/63#issuecomment-282721068

petebacondarwin committed:
build(aio): compute `{@link ...} titles more effectively (#16813)

If a usage of `{@link ...}` does not provide a title then
compute it based on the `title` and/or `name` properties
or set the link to invalid.

Closes #16811

chuckjaz committed:
ci: update build to use TypeScript 2.3.2 (#16707)

vikerman committed:
fix(platform-server): wait for async app initializers to complete before removing server side styles (#16712)

This fixes a flicker when transitioning from server rendered page to client rendered page in lazy loaded routes by waiting for the lazy loaded route to finish loading, assuming initialNavigation on the route is set to 'enabled'.

Fixes #15716

vicb committed:
build: do not create the bundles when updating the public API

- 1.7x faster on my machine (2.7 vs 4.6 min)
- should solve the memory issue

wardbell committed:
feat(aio): buttons for TOC "Contents" label

- Use buttons for the TOC “Contents” label when embedded-and-expandable or TOC on the right to satisfy a11y.
- Add aria-pressed setting for the toggles in TOC and NavItem.
- Clicking the right panel TOC “Contents” button scrolls to top.
- When embedded use same rotating caret as sidebar
- When embedded and no secondaries, “Content” is just a label.
- Gray background for focused buttons rather than outline because can’t get carets to work with outline.

chuckjaz committed:
fix(compiler-cli): allow '==' to compare nullable types (#16731)

Fixes: #16729

* fix(compiler-cli): diagnose issues in conditional of ternary

Fixes: #16730

gkalpak committed:
build(aio): enable HTTP/2 on the preview server (#16826)

Fixes #16780

gkalpak committed:
docs(aio): minor docs fixes for `aio-builds-setup/`

gkalpak committed:
build(aio): fix syntax error in preview server's Dockerfile

gkalpak committed:
build(aio): add script for updating the preview server

petebacondarwin committed:
feat(aio): update aio-shell class when sidenav opens or closes

Foxandxss committed:
build(aio): reduce images weight

matsko committed:
perf(animations): reduce size of animations bundle

before:
animations.umd.js = 33k
animations.umd.min.js = 9.0k

after:
animations.umd.js = 31k
animations.umd.min.js = 7.9k

vicb committed:
refactor(router): don't run the change detection every time an outlet is activated

gkalpak committed:
docs(aio): rework of the upgrade guide

This commit was worked on by a number of people including
@filipesilva, @gkalpak and @wardbell. It contains changes that:

* remove unused files,
* fix the bootstrap approach to ensure that bootstrap is in the correct Zone
* fix unclosed code-example tags
* replace use of "we" with "you"
* remove broken dual router example

Related to angular/angular.io#3541

gkalpak committed:
build(aio): minor `update-preview-server.sh` improvements

petebacondarwin committed:
fix(aio): fix up typing errors in tests

The new version of Jasmine types identified issues with our unit tests.

petebacondarwin committed:
build(aio): do not ignore `is-on-https` in lighthouse check

Lighthouse v1.6.5 treats localhost/1.2.7.0.0.1 as secure domains (i.e. as if they where HTTPS), so we need to stop handling the is-on-https audit specially.

petebacondarwin committed:
fix(aio): add small delay before autoscrolling

This delay allows any async layout to complete
before we autoscroll to a heading anchor.

Closes #16242

matsko committed:
revert: refactor(router): don't run the change detection every time an outlet is activated

This reverts commit 198edb31093ffdecfd13aed17967fb02e650cffe.

matsko committed:
revert: refactor(router): cleanup, simplification

This reverts commit 44d48d9d7aa10bbf03c412ce7e59821c748b5d2d.

matsko committed:
fix(animations): make sure reuseable animation subtitutions work without default params (#16875)

petebacondarwin committed:
fix(aio): ensure that subsections clear floats correctly

petebacondarwin committed:
docs(aio): tidy up docs that use float clearing

Closes #16681

mary-poppins commented:
The angular.io preview for 7915faabbcc98bde46ac3da4fe78f3eb88640e2b is available [here][1].

[1]: https://pr14915-7915faabbcc98bde46ac3da4fe78f3eb88640e2b.ngbuilds.io/

chuckjaz commented:
Obsolete
